### PR TITLE
add quantize_input to QAT quantizer

### PR DESCRIPTION
**Pull request by chenbohua3**

### CODE REVIEW

File: examples/model_compress/quantization/QAT_torch_quantizer.py [hunk 1]
```diff
@@ -47,30 +48,44 @@ def main():
...
+        }, {
+            'quant_types': ['output'],
+            'quant_bits': {'output': 8},
+            'op_names': ['relu2']
+        }, {
+            'quant_types': ['output', 'weight', 'input'],
+            'quant_bits': {'output': 8, 'weight': 8, 'input': 8},
+            'op_names': ['fc1'],
+        }, {
+            'quant_types': ['output', 'weight', ],
```
Comment (J-shang) [inline examples/model_compress/quantization/QAT_torch_quantizer.py:74]: `['output', 'weight', 'input']`?

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 2]
```diff
@@ -522,8 +544,9 @@ def quantize_output(self, output, wrapper, **kwargs):
                                                                        module.ema_decay)
             module.tracked_max_output = update_ema(module.tracked_max_output, current_max,
                                                                        module.ema_decay)
-            module.scale, module.zero_point = update_quantization_param(
-                output_bits, module.tracked_min_output, module.tracked_max_output)
+        module.scale, module.zero_point = update_quantization_param(
```
Comment (J-shang) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:547]: a little confused that we use the same var `module.scale, module.zero_point` in quantizing weight, input, output

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 3]
```diff
@@ -475,7 +478,7 @@ def quantize_weight(self, wrapper, **kwargs):
         config = wrapper.config
         module = wrapper.module
         input = kwargs['input_tensor']  # pylint: disable=redefined-builtin
-        weight = copy.deepcopy(wrapper.module.old_weight.data)
+        weight = wrapper.module.weight
```
Comment (J-shang) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:481]: I am not sure, will this cause some backward issue? And other quantizers also use `copy.deepcopy(wrapper.module.old_weight.data)`, should they also be modified?

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 4]
```diff
@@ -577,6 +598,10 @@ def export_model(self, model_path, calibration_path=None, onnx_path=None, input_
                         module.register_parameter('bias', actual_bias)
                     else:
                         setattr(module, 'bias', None)
+            if hasattr(module, 'input_bit'):
+                calibration_config[name]['input_bit'] = int(module.input_bit)
```
Comment (linbinskn) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:602]: `input_bit` -> `input_bits`

File: nni/algorithms/compression/pytorch/quantization/quantizers.py [hunk 5]
```diff
@@ -503,6 +500,31 @@ def quantize_weight(self, wrapper, **kwargs):
         wrapper.module.weight = weight
         return weight
 
+    def quantize_input(self, inputs, wrapper, **kwargs):
+        config = wrapper.config
+        module = wrapper.module
+        input_bits = get_bits_length(config, 'input')
+        module.input_bit = torch.Tensor([input_bits])
```
Comment (linbinskn) [inline nni/algorithms/compression/pytorch/quantization/quantizers.py:507]: `input_bit`->`input_bits`

**Comment by linbinskn** [comment]
Suggest adding unit test for `quantize_input` in QAT.

**Comment by QuanluZhang** [comment]
@chenbohua3 thanks for this pr, but could add descriptions to explain the reasons and the scenario that `quantize_input` is needed

File: nni/algorithms/compression/pytorch/quantization/quantizers.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:547]: seems indeed have problem here, at least error-prone. and we should think about what kind of information should we export, or how to easily get scale / zero_point from user's point of view

File: examples/model_compress/quantization/QAT_torch_quantizer.py (hunk 1 above)
Comment (chenbohua3) [reply examples/model_compress/quantization/QAT_torch_quantizer.py]: done

File: nni/algorithms/compression/pytorch/quantization/quantizers.py (hunk 2 above)
Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:547]: Yes, it is better to define stand-alone scale & zero_point for each quant_type. I will upload a refactor pr about this when this pr is merged.

File: examples/model_compress/quantization/QAT_torch_quantizer.py [hunk 6]
```diff
@@ -47,30 +48,44 @@ def main():
...
-        'op_types':['Conv2d', 'Linear']
-    }, {
-        'quant_types': ['output'],
-        'quant_bits': 8,
-        'quant_start_step': 1000,
-        'op_types':['ReLU6']
-    }]
+            'quant_types': ['weight', 'input'],
+            'quant_bits': {'weight': 8, 'input': 8},
+            'op_names': ['conv1']
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply examples/model_compress/quantization/QAT_torch_quantizer.py]: done

File: examples/model_compress/quantization/QAT_torch_quantizer.py [hunk 7]
```diff
@@ -47,30 +48,44 @@ def main():
         datasets.MNIST('data', train=False, transform=trans),
         batch_size=1000, shuffle=True)
 
-    model = NaiveModel()
-    '''you can change this to DoReFaQuantizer to implement it
-    DoReFaQuantizer(configure_list).compress(model)
-    '''
+    # conv-bn-relu will be fused into one op. So for this pattern
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply examples/model_compress/quantization/QAT_torch_quantizer.py]: done

File: examples/model_compress/quantization/QAT_torch_quantizer.py [hunk 8]
```diff
@@ -47,30 +48,44 @@ def main():
...
+            'quant_types': ['output', 'weight', ],
+            'quant_bits': {'output': 8, 'weight': 8, 'input': 8},
+            'op_names': ['fc2'],
+        }]
+
+    model = NaiveModel().to(device)
+    dummy_input = torch.randn(1, 1, 28, 28).to(device)
     optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.5)
-    quantizer = QAT_Quantizer(model, configure_list, optimizer)
+    quantizer = QAT_Quantizer(model, configure_list, optimizer, dummy_input=dummy_input)
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply examples/model_compress/quantization/QAT_torch_quantizer.py:83]: done

File: nni/algorithms/compression/pytorch/quantization/quantizers.py (hunk 3 above)
Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:481]: have rebased the master

File: nni/algorithms/compression/pytorch/quantization/quantizers.py (hunk 5 above)
Comment (chenbohua3) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:507]: done

**Comment by chenbohua3** [comment]
@QuanluZhang Two reasons: 
1. the input of the whole network should be quantized. 
2. Not all layers have its quantization kernel in the backend. Let A be the layer that has corresponding quantization kernel on the backend and B be the layer that does not. If we have the structure like A-B-A, we should configure to quantize the input of the second A.

**Comment by chenbohua3** [comment]
@linbinskn ut is added

**Review by QuanluZhang — APPROVED** [review]

**Review by J-shang — APPROVED** [review]

File: nni/algorithms/compression/pytorch/quantization/quantizers.py (hunk 5 above)
Comment (linbinskn) [reply nni/algorithms/compression/pytorch/quantization/quantizers.py:503]: `module.input_bit` -> `module.input_bits`

**Review by linbinskn — APPROVED** [review]

**Pull request merged**